llpc/lower/llpcSpirvLowerCfgMerges.cpp

```diff
@@ -0,0 +1,608 @@
+// Pass creator, creates the pass of SPIR-V lowering terminator operations
+ModulePass *createLegacySpirvLowerCfgMerges() {
+  return new LegacySpirvLowerCfgMerges();
+}
+
+// =====================================================================================================================
+LegacySpirvLowerCfgMerges::LegacySpirvLowerCfgMerges() : ModulePass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on (empty on entry)
+bool LegacySpirvLowerCfgMerges::runOnModule(Module &module) {
+  return Impl.runImpl(module);
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on (empty on entry)
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+PreservedAnalyses SpirvLowerCfgMerges::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  bool changed = runImpl(module);
+  // In practice there are unlikely to be any analyses this early, but report accurate status anyway.
+  return changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+}
+
+class OpPrinter {
+  BasicBlock *m_bb;
+  Function *m_func;
+
+public:
+  OpPrinter(BasicBlock *BB) : m_bb(BB), m_func(nullptr) {}
+  OpPrinter(Function *F) : m_bb(nullptr), m_func(F) {}
+
+  void print(raw_ostream &OS) const {
+    if (m_bb) {
+      m_bb->printAsOperand(OS);
+      OS << " (" << m_bb << ")";
+    }
+    if (m_func) {
+      m_func->printAsOperand(OS);
+      OS << " (" << m_func << ")";
+    }
+  }
+};
+
+raw_ostream &operator<<(raw_ostream &OS, const OpPrinter &V) {
+  V.print(OS);
+  return OS;
+}
+
+struct LoopDesc {
+  LoopDesc()
+      : loopHeader(nullptr), mergeBlock(nullptr), continueBlock(nullptr), backedgeBlock(nullptr), sigmaBlock(nullptr),
+        returnPhi(nullptr), returnValuePhi(nullptr), parent(nullptr), function(nullptr), depth(-1),
+        convergentOps(false){};
+
+  int computeDepth() {
+    if (depth < 0) {
+      if (parent)
+        depth = parent->computeDepth() + 1;
+      else
+        depth = 0;
+    }
+    return depth;
+  }
+
+  BasicBlock *loopHeader;
+  BasicBlock *mergeBlock;
+  BasicBlock *continueBlock;
+  BasicBlock *backedgeBlock;
+  BasicBlock *sigmaBlock;
+  PHINode *returnPhi;
+  PHINode *returnValuePhi;
+  LoopDesc *parent;
+  Function *function;
+  int depth;
+  bool convergentOps;
+  SmallVector<BasicBlock *, 2> returnBlocks;
+};
+
+static bool isConvergentOp(const Instruction *inst) {
+  const CallInst *callInst = dyn_cast<const CallInst>(inst);
+  if (!callInst)
+    return false;
+
+  auto callee = callInst->getCalledFunction();
+  return callee ? callee->isConvergent() : false;
```

Comment:
callInst->isConvergent() should just work? It will check against the function being called.

Reply:
Handling for this has changed so this should be fine now.